discovery/manager.go

```diff
@@ -296,6 +296,49 @@ func (m *Manager) startProvider(ctx context.Context, p *Provider) {
 
 // cleaner cleans resources associated with provider.
 func (m *Manager) cleaner(p *Provider) {
+
+	// Before deleting the targets of p from m.targets definitely, we should ask
+	// the consumers to forget about them
+
+	// this logic should go somewhere else as this will block the reload
+	// maybe add a m.droppedTargets and make sender() expire them?
+	// maybe make allGroups to that? as it keeps empty targets
+
+	tSets := map[string][]*targetgroup.Group{}
+	m.targetsMtx.Lock()
+	p.mu.RLock()
+	for pkey, tsets := range m.targets {
+		// TODO: m.subs can at most contain one sub, don't see wht the plural and why a map? (maybe easy to use?)
```

Comment:
am I correct?

Reply:
There can be multiple subs. if  2 scrape configs share the same service discovery but not the same relabel config.

ts=2023-11-17T09:42:41.571Z caller=manager.go:194 level=debug component="discovery manager scrape" msg="Starting provider" provider=static/0 subs="[scrape_config_1 scrape_config_2]"

```yaml
scrape_configs:
  # First scrape configuration
  - job_name: 'scrape_config_1'
    static_configs:
      - targets: ['localhost:9090']  # Define the targets here. This is the shared part.
    relabel_configs:
      - source_labels: [__address__]
        regex: '(.+):9090'
        target_label: '__param_target'
        replacement: '${1}:9091'  # Relabeling for the first config.

  # Second scrape configuration
  - job_name: 'scrape_config_2'
    static_configs:
      - targets: ['localhost:9090']  # Same targets as the first config.
    relabel_configs:
      - source_labels: [__address__]
        regex: '(.+):9090'
        target_label: 'instance'
        replacement: '${1}'  # Different relabeling for the second config.

```

Reply:
Thanks

TBH, I haven't do any tests for that, I just looked at the code, but I was missing

https://github.com/prometheus/prometheus/blob/1b84c01b7626cec7f10df5ca17a40da267364f15/discovery/manager.go#L421-L424

Where we can insert multiple subscribers, my bad.

with the new manager (your log is from the legacy one I think), we can see that with a simpler example (that doesn't really make sense) where the configs are equal:

```
scrape_configs:
  - job_name: 'example1'
    static_configs:
      - targets: ['localhost:8000']
  - job_name: 'example2'
    static_configs:
      - targets: ['localhost:8000']
```

(I'll add a unit test for that case (two jobs sharing the same SD file config) to make sure the fix doesn't mess up with the other job's targets)

Reply:
I'll need to investigate the use of the same `refTargets` for all new subs here: https://github.com/prometheus/prometheus/blob/59844498f7b12f16c7f004aa951bbb14cdb83991/discovery/manager.go#L222-L245